### PR TITLE
feat: Add deterministic memory-file drift check (tools/memcheck)

### LINKED ISSUE #1053: tools: add a memory-file freshness check (CLAUDE.md vs reality) to the drift harness
## Problem

The repo ships a sophisticated drift-detection framework — `.claude/drift-allowlist.yaml` (+ schema), `tools/driftcheck` (#673), and five `check-*-drift` skills covering code ↔ schema ↔ docs ↔ examples ↔ spec (`.claude/skills/check-suite/README.md`) — but applies **none** of it to its own `CLAUDE.md` memory files, even though those files make dozens of concrete, checkable assertions that can rot.

This is the repo's own **dogfooding gap**: it lints docs↔schema parity in CI but not memory-file↔reality parity. "Stale instructions are worse than missing instructions" — the agent confidently applies dead patterns. Two already-live examples:

1. **The pinning-rule heading** in `.github/CLAUDE.md:3` mandates the inverse of the repo's actual SHA-pin convention (filed separately) — a trivial command/path-coverage check would have flagged it.
2. **A stale path in the root memory file:** `CLAUDE.md:90` cites `testdata/gedcom/shakespeare.ged`, but no file resolves at that repo-relative path — the real location is `glx/testdata/gedcom/5.5.1/shakespeare-family/shakespeare.ged`. A path-existence check catches this immediately.

## Proposed check

Add a lightweight, deterministic, **offline** (API-key-free) `check-memory-drift` — a skill or a `make lint` step reusing the `tools/driftcheck` pattern (#673) — that, for every `CLAUDE.md` (and `AGENTS.md` if adopted), asserts:

1. Every `make <target>` referenced exists in the `Makefile`. *(Verified present today: `build`, `test`, `lint`, `check-schemas`, `check-links`, `docs-cli`.)*
2. Every repo-relative file/dir path referenced exists on disk. *(Catches the `shakespeare.ged` drift above.)*
3. The import path `github.com/genealogix/glx/go-glx` matches `go.mod`'s module line.
4. (Optional) Issue/PR references in instructions resolve.

These are cheap invariants and mirror the "command-coverage / freshness" checks emerging in 2026 agent-memory tooling. Pair with a stripped `Last reviewed:` HTML-comment anchor per file so freshness costs zero agent context.

## Why it's distinct from the existing harness

`check-suite/README.md`'s migration table enumerates docs/schema/code/example/spec drift checks (#298/#309/#311/#839/#795/#910); **none** targets the `CLAUDE.md`/`AGENTS.md` instruction files. No backlog issue covers memory-file drift detection.

## Cross-refs

- Epic #1016 (consolidate the `check-*` drift suite) — natural home for a new surface.
- #673 (offline `tools/driftcheck`), #797 (drift allowlist) — the patterns to reuse.
- Companion issues: pinning-heading drift (the motivating example), AGENTS.md adoption (adds a file this check should cover).

### PR DESCRIPTION
**Pull request by noahbjohnson**
## What and why

The repo lints docs↔schema parity in CI via the `check-*` drift suite + `tools/driftcheck`, but applies **none** of it to its own `CLAUDE.md` memory files — even though those files make dozens of concrete, checkable assertions that rot. Stale instructions are worse than missing ones: an agent confidently applies dead patterns. This closes that dogfooding gap (#1053).

`tools/memcheck` is an offline, API-key-free counterpart to the LLM-based suite. For every `CLAUDE.md` (and `AGENTS.md` if adopted) it asserts three cheap, deterministic invariants:

1. **Make targets exist** — every `` `make <target>` `` referenced (inline, or in command position inside a fenced block) is a real `Makefile` target.
2. **Paths exist** — every concrete file/dir path resolves on disk, checked relative to the repo root **and** the memory file's own directory.
3. **Import path matches go.mod** — any `github.com/<org>/<repo>` import path under the module's org is prefixed by the `go.mod` module line (catches a module rename the docs didn't track).

The hard part is separating a genuine repo path from path-shaped look-alikes. The rule is syntactic and conservative (it would rather miss an obscure reference than gate CI on a false one): a **file claim** is a token with a slash *and* a known extension; a **directory claim** is a multi-segment trailing-slash token. This excludes GitHub actions (`actions/checkout`), stdlib imports (`os/exec`), build artifacts (`bin/glx`), routes (`/cli`), home paths (`~/.config/...`), branch prefixes (`claude/`), sibling-org GitHub URLs, and single-segment basenames. Intentional exceptions live in `.claude/memory-drift-allowlist.yaml`.

It runs in `make check-memory-drift`, is included in `make check`, and gets a dedicated CI job in `validate-spec.yml` alongside `check-code-drift`. It surfaced and fixed two stale fixture paths in the root `CLAUDE.md` (`testdata/gedcom/shakespeare.ged` → `glx/testdata/gedcom/5.5.1/shakespeare-family/shakespeare.ged`, and the `minimal-70.ged` path).

Folded in (#854): the `/compact-changelog` command was the only command in `.claude/` without `allowed-tools`/`model` frontmatter. It now declares a minimal tool surface (covering the `git`/`awk`/`grep`/`head`/`sort`/`comm`/`printf` it actually runs), pins `model: claude-opus-4-8` to match its siblings, sets `disable-model-invocation: true` (it mutates a tracked file), and adds an `argument-hint`. Scoped to frontmatter only — the JSON-output/idempotency/provenance work from #856 is intentionally **not** included.

## Related issues

Closes #1053
Refs #854
Refs #1016

## Review focus

- **False-positive safety of the path/import discriminator** (`tools/memcheck/scan.go`, `classifyPathClaim` / `importFindings`) — this is the crux; a false positive here gates CI. The change went through an adversarial review pass that empirically caught and fixed `~/`, quoted-token, sibling-org-URL, and underscore-target edge cases.
- Whether the deterministic-only scope is right: issue/PR-reference resolution (needs network) and the optional `Last reviewed:` freshness marker are intentionally deferred.

## Testing

- `go test ./tools/memcheck/` — passing, 86.6% statement coverage (table-driven discriminator tests + a `TestRealTreeHasNoDrift` guard that runs under `make test`).
- `golangci-lint run ./tools/memcheck/...` — clean. `go vet ./...` — clean.
- `make check-memory-drift` — clean (exit 0) against the real tree.

## Breaking changes

None. New developer tooling + a new (passing) CI job.

### CODE REVIEW

File: tools/memcheck/scan.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+	if strings.HasPrefix(tok, "-") || strings.HasPrefix(tok, "/") || strings.HasPrefix(tok, "~") {
+		return "", false, false // flag, absolute path / route, or home-relative path
+	}
+	if strings.Contains(tok, "://") || strings.ContainsAny(tok, "@:=") {
+		return "", false, false // URL, ref@version, key:value, assignment
+	}
+	if strings.HasPrefix(tok, "github.com/") {
+		return "", false, false // import path, handled by importFindings
+	}
+
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/scan.go]: Addressed in e423117c. `classifyPathClaim` now rejects any token containing a `..` path segment (via `slices.Contains(strings.Split(tok, "/"), "..")`), so memcheck never resolves a parent traversal — added `../secrets.txt`, `go-glx/../../etc/passwd.txt`, and `docs/../../../tmp/` to the discriminator test table.

File: tools/memcheck/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+- Tokens with a leading `/`, a `://`, an `@`, a `:` / `=`, glob/shell
+  metacharacters, or whitespace are never treated as paths.
+- Only inline code spans are scanned for paths; fenced code blocks (command/code
+  examples) are scanned only for `make` invocations.
+
+## Allowlist
+
+A deliberately-documented reference that should *not* resolve (an illustrative
+example path, or a `make` target inside a fenced shell comment) can be suppressed
+by adding an entry to `.claude/memory-drift-allowlist.yaml`:
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/README.md]: Addressed in e423117c. The README example was misleading — the scanner does not flag `make` targets inside fenced shell comments. Reworded the suppressible-case example to "a `make` target that is documented before it exists in the Makefile", which is a case the scanner actually emits.

File: tools/memcheck/scan.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+	makeRe = regexp.MustCompile(`\bmake\s+([A-Za-z][A-Za-z0-9_-]{2,})`)
+	// makeLineRe matches a `make <target>` invocation in command position at the
+	// start of a fenced-block line (after optional indentation and a `$`/`>`
+	// prompt). Anchoring here keeps a shell comment such as `# make sure to
+	// rebuild` from being read as the target "sure".
+	makeLineRe = regexp.MustCompile(`^\s*(?:[$>]\s+)?make\s+([A-Za-z][A-Za-z0-9_-]{2,})`)
+	// importRe matches a github.com import/module path. The trailing class
+	// excludes '@' and backslash so version suffixes and regex-escaped doc
+	// strings terminate the match cleanly.
+	importRe = regexp.MustCompile(`github\.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_./-]+`)
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/scan.go]: Addressed in e423117c. Anchored `importRe` with a leading `\b` so a longer hostname (e.g. `notgithub.com/org/repo`) can no longer match on its `github.com/...` suffix. Added a `notgithub.com/...` regression case to `TestImportFindings` (expects 0 findings).

File: tools/memcheck/scan.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+	// start of a fenced-block line (after optional indentation and a `$`/`>`
+	// prompt). Anchoring here keeps a shell comment such as `# make sure to
+	// rebuild` from being read as the target "sure".
+	makeLineRe = regexp.MustCompile(`^\s*(?:[$>]\s+)?make\s+([A-Za-z][A-Za-z0-9_-]{2,})`)
+	// importRe matches a github.com import/module path. The leading `\b` anchors
+	// the host so a longer hostname (e.g. `notgithub.com/org/repo`) can't match on
+	// its `github.com/...` suffix. The trailing class excludes '@' and backslash
+	// so version suffixes and regex-escaped doc strings terminate the match
+	// cleanly.
+	importRe = regexp.MustCompile(`\bgithub\.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_./-]+`)
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/scan.go]: Addressed in e9029414. The earlier `\b` anchor was insufficient — `evil.github.com/...` has a word boundary before `github` (preceded by `.`), so it still matched the `github.com/...` suffix. `importRe` now anchors the host with `(?:^|[^A-Za-z0-9.-])` and captures the path in group 1, so neither a prefix label (`notgithub.com/...`) nor a subdomain (`evil.github.com/...`) can match; `https://github.com/...` still matches and is handled by the existing `://` guard. This also closed a real false positive in `importFindings`. Added a subdomain regression case to `TestImportFindings`.

File: tools/memcheck/main.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+	flag.BoolVar(&o.verbose, "v", false, "list allowlist-suppressed findings")
+	flag.Parse()
+
+	return o
+}
+
+// run performs the check and returns the process exit code (0 = clean,
+// 1 = gating drift found). It writes the human-readable report to out; an
+// operational failure is returned as an error and mapped to exit 2 by realMain.
+func run(opts options, out io.Writer) (int, error) {
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/main.go]: Addressed in bb1d7eec. The `run` doc comment now documents all three exit codes (0 = clean, 1 = gating drift, 2 = operational failure) and notes that code 2 is returned alongside a non-nil error that `realMain` prints before exiting 2. Kept the `2, err` returns rather than removing them, since `run` is the single source of the exit code and tests assert on it directly.

File: tools/memcheck/scan.go [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+			start := loc[2] // group 1: the github.com path, sans any boundary char
+			tok := strings.TrimRight(line[start:loc[3]], "./-")
+			if !strings.HasPrefix(tok, r.orgPrefix) {
+				continue // third-party (or different-org) import: not our concern
+			}
+			// A web URL to a same-org repo (`https://github.com/genealogix/...`)
+			// is a link, not a Go import — leave it alone.
+			if start >= len("://") && line[start-len("://"):start] == "://" {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/scan.go]: Good catch — addressed in 686623ff. The web-URL guard previously only matched when the three characters immediately before the `github.com` match were `://`, so `ssh://git@github.com/<org>/<repo>` (match preceded by `git@`) bypassed it and could be flagged as import drift. Replaced the fixed-offset check with a new `withinSchemeToken(line, start)` helper that treats the match as a URL when a `://` appears anywhere in its whitespace-delimited token — covering `https://`, `ssh://git@…`, and any other `scheme://…` form. Added an `ssh://git@github.com/…` regression case to `TestImportFindings` (expects 0 findings).

File: .github/workflows/validate-spec.yml [hunk 7]
```diff
@@ -57,6 +57,22 @@ jobs:
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    steps:
+      - uses: actions/checkout@v6
+
+      - name: Set up Go
+        uses: actions/setup-go@v6
+        with:
+          go-version: '1.26'
+          cache: true
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply .github/workflows/validate-spec.yml]: Addressed in e4c18cc3 — aligned the outlier to `actions/checkout@v7`.

On the pinning policy: `.github/CLAUDE.md:3-23` is specifically about **third-party** actions (e.g. `ossf/scorecard-action`, `sigstore/cosign-installer`) that have repeatedly broken because their maintainers don't publish a floating `@vN` major tag. First-party `actions/*` reliably publish floating majors, and this repo pins them that way everywhere — `actions/checkout@v7` (25 uses) and `actions/setup-go@v6` (11 uses). My job's `actions/checkout@v6` was just inconsistent with that convention, not a third-party supply-chain risk; `setup-go@v6` already matched. Pinning this one job to a SHA/patch would make it the lone outlier in the opposite direction, so I matched the established `@v7`/`@v6` convention instead.

File: tools/memcheck/report.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	fmt.Fprintf(&b, "memcheck: checked %d memory file(s) (CLAUDE.md/AGENTS.md) against the repository\n", memFiles)
+
+	if verbose && len(suppressed) > 0 {
+		fmt.Fprintf(&b, "\nSuppressed by .claude/memory-drift-allowlist.yaml (%d):\n", len(suppressed))
+		for i := range suppressed {
+			s := &suppressed[i]
+			fmt.Fprintf(&b, "  - %s:%d [%s] %s — %s\n",
+				s.finding.File, s.finding.Line, s.finding.Kind, s.finding.Token, s.entry.Reason)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/report.go:39]: Addressed in 73a152de. `buildReport` now takes the effective allowlist path (`opts.allowlistFile`, default `.claude/memory-drift-allowlist.yaml`) and prints it in both the verbose "Suppressed by …" header and this footer, so a run with `-allowlist <other>` points users at the file actually in use. Added `TestBuildReportEffectiveAllowlistPath`, which asserts a custom value is cited in both places and that the default path no longer leaks.

File: tools/memcheck/report.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+	fmt.Fprintf(&b, "\nSummary: %d major, %d minor, %d info",
+		counts[severityMajor], counts[severityMinor], counts[severityInfo])
+	if len(suppressed) > 0 {
+		fmt.Fprintf(&b, " (%d suppressed by allowlist)", len(suppressed))
+	}
+	b.WriteString("\n\nThe repository is the source of truth — update the memory file to match,\n" +
+		"or add a triaged entry to .claude/memory-drift-allowlist.yaml if the\n" +
+		"reference is an intentional, non-existent example.\n")
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply tools/memcheck/report.go]: Addressed in 73a152de. `buildReport` now takes the effective allowlist path (`opts.allowlistFile`, default `.claude/memory-drift-allowlist.yaml`) and prints it in both the verbose "Suppressed by …" header and this footer, so a run with `-allowlist <other>` points users at the file actually in use. Added `TestBuildReportEffectiveAllowlistPath`, which asserts a custom value is cited in both places and that the default path no longer leaks.

**Review by isaacschepp — APPROVED** [review]
